Modify hello to be an enumration and add a backend

cminyard opened the pull request:
This is a set of changes to make the "hello" example a little more useful.  It shows a lot more of the things I expect you would be using when using clixon.

I don't think this is perfect, and I have a lot of questions still.  But something like this would be extremely helpful for new users.

cminyard committed:
hello: Add a backend file that prints out info

Add a backend, this will print out information about the transaction,
but that's it.

Signed-off-by: Corey Minyard <cminyard@mvista.com>

cminyard committed:
hello: Extend the backend

Add validation of the data, keep a local state and return the proper
state information.

Signed-off-by: Corey Minyard <minyard@acm.org>

cminyard committed:
hello: use a file for the backend

Use the existence of a file for the status for the hello world backend.

Signed-off-by: Corey Minyard <minyard@acm.org>

cminyard committed:
hello: Change the value to an enumeration

This lets us demonstrate changing the value.  The /tmp/world file
contains a "to" value.

Signed-off-by: Corey Minyard <minyard@acm.org>

cminyard committed:
hello: Update the version of the hello yang file

Because it's changed...

Signed-off-by: Corey Minyard <minyard@acm.org>

cminyard committed:
hello: Fix the yang filename

Make the versions match.

Signed-off-by: Corey Minyard <minyard@acm.org>

cminyard committed:
hello: Add printing of the full xml source and target trees

To see what's in them.  The standard functions don't print the prefix or
flags.

Signed-off-by: Corey Minyard <minyard@acm.org>

olofhagsand commented:
Nice initiative!

cminyard committed:
hello: Clean up xml print

Handle body nodes more cleanly.

Signed-off-by: Corey Minyard <minyard@acm.org>

olofhagsand commented on hello/src/hello_backend_plugin.c:
```diff
@@ -196,6 +256,11 @@ hello_validate(clicon_handle h, transaction_data td) {
     int rv;
 
     printf("*****hello validate*****\n");
+    printf("src:\n");
```
This will show up on the backend's stdout. This is limiting to a debug case where you run the backend in the foreground.
I would add that as a log instead so you could steer the prints to syslog/stderr/stdout as described here: https://clixon-docs.readthedocs.io/en/latest/backend.html#command-line-options.
Logging XML can be done with the macro  clixon_log_xml(). Or debug with  clixon_debug_xml().

olofhagsand commented on hello/src/hello_backend_plugin.c:
```diff
@@ -54,8 +54,16 @@
 
-static bool hello_state = false;
+/*
+ * This file reflects the state of hello world.  If it exists, hello
```
I am a little unsure of the purpose for the file. Does it represent "system state" not present in the config?

olofhagsand commented on hello/src/hello_backend_plugin.c:
```diff
@@ -79,29 +87,46 @@ hello_begin(clicon_handle h, transaction_data td) {
     if (strcmp(xml_name(vec), "hello") != 0)
 	return 0;
+    /*
+     * It would be nice to be able to use nscache_xxx() for this.
```
The cache is built on-demand. Yes, you should be able to use such a function. It is not so nice to have this in the example code.
Maybe we could add such a function in the lib if there is nothing appropriate.
The function I think could be used here is xml2ns()

olofhagsand commented on hello/src/clixon-hello@2019-04-17.yang:
```diff
@@ -2,7 +2,7 @@ module clixon-hello {
     yang-version 1.1;
     namespace "urn:example:hello";
     prefix he;
-    revision 2019-04-17 {
```
According to https://datatracker.ietf.org/doc/html/rfc7950#section-7.1.9 new revisions should be added before existing.

olofhagsand reviewed (commented):
Thanks for improving the hello application.
Could you please add a README to describe the purpose and usage?
Your additions seem to be a better illustration of the commit/transaction semantics?
I could still see the use of a minimal application just to exercise the config+build system (the previous hello was not that either).
Could you also please have a look at  https://github.com/clicon/clixon/blob/master/CONTRIBUTING.md#c-style especially in an example its good to be normative in style.

cminyard committed:
hello: Clean up the hello world example changes

Add a lot of comments, replace hand-coded xml functions with standard
one, use clixon_debug() instead of printf in a lot of places, and
comment out the printing of the XML trees by default.

Also show an example of both parsing the source and target trees or
parsing the add,delete,change vectors.

Signed-off-by: Corey Minyard <minyard@acm.org>

cminyard commented:
I updated this to address the comments and to add even more useful information and examples.

cminyard committed:
Fix the revision of the yang module

You aren't supposed to delete the old ones, just add new ones.

Signed-off-by: Corey Minyard <corey@minyard.net>
